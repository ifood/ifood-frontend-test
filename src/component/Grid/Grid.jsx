import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const GridDiv = styled.div`
  display: grid;
  grid-template-columns: repeat(
    auto-fill,
    minmax(${(props) => props.columnMinWidth}, 1fr)
  );
  gap: 1em;
`;

function Grid({ title, items, columnMinWidth = '200px', itemRenderer }) {
  if (items == null || items.length === 0) {
    return null;
  }

  function renderItem(item) {
    if (itemRenderer != null) {
      return itemRenderer(item);
    }
    return null;
  }

  return (
    <>
      {title != null && <h1>{title}</h1>}
      <GridDiv columnMinWidth={columnMinWidth}>{items.map(renderItem)}</GridDiv>
    </>
  );
}

Grid.propTypes = {
  title: PropTypes.string,
  items: PropTypes.array,
  columnMinWidth: PropTypes.string,
  itemRenderer: PropTypes.func.isRequired,
};

export default React.memo(Grid);
