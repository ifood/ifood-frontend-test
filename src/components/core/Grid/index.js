import styled, { css } from "styled-components";

import authBackground from "assets/images/auth-background.jpg";

const authScreenStyle = css`
  background: url(${authBackground}) center no-repeat;
  background-size: cover;

  &:after {
    content: "";
    position: absolute;
    left: 0;
    right: 0;
    height: 100%;
    width: 100%;
    z-index: 1;
    background: rgb(0 0 0 / 81%);
  }
`;

export const Main = styled.main`
  width: 100%;
`;

export const Container = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  position: relative;
  height: 100%;
  min-height: 100vh;
  padding: 5rem 10rem;

  ${({ authScreen }) => !authScreen && authScreenStyle};
`;

export const Wrapper = styled.div`
  position: relative;
  width: 100%;
  z-index: 2;
`;

export const Div = styled.div`
  width: 100%;
  height: auto;
`;

export const FlexCenter = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;

  flex-direction: ${({ row }) => (row ? "row" : "column")};
`;

export const Columns = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(350px, 1fr));
  grid-column-gap: 6rem;
  grid-row-gap: 6rem;
`;

export const CustomColumns = styled.div`
  display: grid;
  grid-template-columns: repeat(
    auto-fit,
    minmax(${({ minmax }) => minmax}, 1fr)
  );
  grid-column-gap: 2rem;
  grid-row-gap: 2rem;
`;
