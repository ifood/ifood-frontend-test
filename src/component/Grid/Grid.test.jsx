import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Grid from './Grid';

const items = [
  { id: 1, description: 'Item 1' },
  { id: 2, description: 'Item 2' },
];

describe('<Grid />', () => {
  it('snapshot', () => {
    const grid = mount(
      <Grid title="Grid title" items={items} itemRenderer={() => {}} />
    );
    jestExpect(toJson(grid)).toMatchSnapshot();
  });

  it('should return null if items are not provided', () => {
    const grid = shallow(<Grid />);
    expect(grid.type()).to.be.null;
  });

  it('should not render a h1 if title are not provided', () => {
    const grid = shallow(<Grid items={items} itemRenderer={() => {}} />);
    expect(grid.find('h1')).to.have.lengthOf(0);
  });

  it('should call itemRenderer for each item', () => {
    const itemRenderer = jest.fn();
    shallow(<Grid items={items} itemRenderer={itemRenderer} />);
    expect(itemRenderer.mock.calls).to.have.lengthOf(items.length);
  });
});
