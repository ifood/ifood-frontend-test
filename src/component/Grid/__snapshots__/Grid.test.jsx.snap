// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> snapshot 1`] = `
<Memo(Grid)
  itemRenderer={[Function]}
  items={
    Array [
      Object {
        "description": "Item 1",
        "id": 1,
      },
      Object {
        "description": "Item 2",
        "id": 2,
      },
    ]
  }
  title="Grid title"
>
  <h1>
    Grid title
  </h1>
  <styled.div
    columnMinWidth="200px"
  >
    <div
      className="sc-AxjAm eeWsGe"
    />
  </styled.div>
</Memo(Grid)>
`;
